02-documentation/developer-guides/building-a-nextjs-blog-with-squidex.md

```diff
@@ -0,0 +1,588 @@
+
+```jsx
+export default function BlogPost() {
+  return <div>Body</div>;
+}
+```
+
+Click one of the links on the homepage to visit the page where your blog posts will be read.
+
+Since you will need to render markdown content on this page, you will install [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote), a package that parses markdown content and renders it in your Next.js application. Stop the development server by pressing `Ctrl/Cmd + C` on your terminal and run the following command to install `next-mdx-remote`:
+
+```bash
+yarn add next-mdx-remote
+```
+
+Restart the development server after installation:
+
+```
+yarn dev
+```
+
+Import `serialize` and `MDXRemote` from `next-mdx-remote` at the top of `[slug].js`:
+
+```jsx
+import { serialize } from "next-mdx-remote/serialize";
+import { MDXRemote } from "next-mdx-remote";
+```
+
+Import `fetchAPI` at the top of `[slug].js` as you will use it to fetch the contents your blog posts from Squidex:
+
+```js
+import fetchAPI from "../lib/squidex";
+```
+
+Export a `getStaticProps` function from `[slug].js`:
+
+```js
+export async function getStaticProps({ params }) {
+  const data = await fetchAPI(
+    `
+  query QueryPosts($query: String!) {
+    queryPostsContents(filter: $query) {
+      id
+      flatData {
+        title
+        content
+      }
+    }
+  }
+  `,
+    {
+      variables: {
+        query: `data/Slug/iv eq '${params.slug}'`,
+      },
+    }
+  );
+
+  if (data.queryPostsContents.length === 0) {
+    return {
+      notFound: true,
+    };
+  }
+  const post = data.queryPostsContents[0].flatData;
+
+  const mdxSource = await serialize(post.content);
+
+  return {
+    props: {
+      post: post,
+      source: mdxSource,
+    },
+  };
+}
+```
+
+This returns data for the blog post associated with a particular slug or returns a 404 error if no such post exists.
+
+At this point, you may receive an error because when you use `getStaticPaths` with dynamic paths in Next.js, you are required to also export a `getStaticPaths` function that tells Next.js [what pages to generate at build time](https://nextjs.org/docs/basic-features/data-fetching#getstaticpaths-static-generation).
+
+Export a `getStaticPaths` function in `squidex-blog/pages/index.js`:
+
+```js
+export async function getStaticPaths() {
+  const data = await fetchAPI(`
+  {
+    queryPostsContents {
+      id,
+      flatData {
+        slug
+      }
+    }
+  }
+  `);
+
+  return {
+    paths: data.queryPostsContents.map((post) => {
+      return {
+        params: {
+          slug: post.flatData.slug,
+        },
+      };
+    }),
+    fallback: false,
+  };
+}
+```
+
+When you save the file, refresh the page to fetch data from Squidex. Right now, again, although the data is available to the page, you cannot see it since we have not rendered the blog post. You will now render the loaded blog post on the page.
```

Comment:
I thought that the development server makes the refresh automatically...

Reply:
What the development server does is that it hot reloads the components in your application. So whenever you make changes to a component, the changes become visible automatically. However, things work a bit differently for `getStaticProps`. The `getStaticProps` function only runs when the page is initially rendered. So any changes you make to the `getStaticProps` function will be applied on the server-side, but there will be no visible changes on the preview page since the `getStaticProps` function is not called because an initial render is not happening. When you refresh the page, you force the initial render to happen again and the changes to the `getStaticProps` become visible.

I have included this detail in the article.